lang/src/org/partiql/lang/ast/passes/StatementRedactor.kt

```diff
@@ -82,238 +73,256 @@ fun redact(statement: String,
+     */
+    fun getRedactedStatement(): String {
+        val lines = statement.lines()
+        val totalCharactersInPreviousLines = IntArray(lines.size)
+        for (lineNum in 1 until lines.size) {
+            totalCharactersInPreviousLines[lineNum] = totalCharactersInPreviousLines[lineNum - 1] + lines[lineNum - 1].length + 1
         }
-    }
 
-    private fun redactOnNAry(node: NAry) {
-        val rewrittenArgs = node.args
+        val redactedStatement = StringBuilder(statement)
+        var offset = 0
+        sourceLocationMetaForRedaction.sortWith(compareBy<SourceLocationMeta> { it.lineNum }.thenBy { it.charOffset })
 
-        when (node.op) {
-            AND, OR -> {
-                rewrittenArgs.forEach { redactOnField(it) }
-            }
-            EQ, NE, GT, GTE, LT, LTE, IN -> {
-                if (rewrittenArgs.size != 2) {
-                    throw IllegalArgumentException(INVALID_NUM_ARGS)
-                }
-                if (!validateSafeFieldNames(rewrittenArgs[0])) {
-                    redactOnField(rewrittenArgs[1])
-                }
-            }
-            ADD, SUB -> {
-                when (rewrittenArgs.size) {
-                    1 -> {
-                        redactOnField(rewrittenArgs[0])
-                    }
-                    2 -> {
-                        redactOnField(rewrittenArgs[0])
-                        redactOnField(rewrittenArgs[1])
-                    }
-                    else -> {
-                        throw IllegalArgumentException(INVALID_NUM_ARGS)
-                    }
-                }
-            }
-            MUL, DIV, MOD -> {
-                if (rewrittenArgs.size != 2) {
-                    throw IllegalArgumentException(INVALID_NUM_ARGS)
-                }
-                redactOnField(rewrittenArgs[0])
-                redactOnField(rewrittenArgs[1])
-            }
-            BETWEEN -> {
-                if (rewrittenArgs.size != 3) {
-                    throw IllegalArgumentException(INVALID_NUM_ARGS)
-                }
-                if (!validateSafeFieldNames(rewrittenArgs[0])) {
-                    redactOnField(rewrittenArgs[1])
-                    redactOnField(rewrittenArgs[2])
-                }
-            }
-            NOT -> {
-                if (rewrittenArgs.size != 1) {
-                    throw IllegalArgumentException(INVALID_NUM_ARGS)
-                }
-                val arg = rewrittenArgs[0]
-                when (arg) {
-                    is NAry -> redactOnNAry(arg)
-                    is Typed -> redactOnTyped(arg)
-                    else -> { /* intentionally blank */ }
-                }
+        sourceLocationMetaForRedaction.map {
+            val length = it.length.toInt()
+            val lineNum = it.lineNum.toInt()
+            if (lineNum < 1 || lineNum > totalCharactersInPreviousLines.size) {
+                throw IllegalArgumentException("$INPUT_AST_STATEMENT_MISMATCH, line number: $lineNum")
```

Comment:
The error message "Input ast should be the parsed result of the input statement" doesn't clearly indicate what went wrong here...  Even reading the code I'm not 100% why this is an error condition and what I should do to solve it.

Reply:
To give some background, there was a `redact` [function call](https://github.com/partiql/partiql-lang-kotlin/blob/migrate-statement-redactor-375/lang/src/org/partiql/lang/ast/passes/StatementRedactor.kt#L88) that allows users to pass in the `statement: String` and the `ast: ExprNode` rather than just the `statement`. I think the idea here was if you already had the parsed `ExprNode` from `statement`, you could perform the redaction logic without having to re-parse. 

This check was here if the `statement` and `ast` were mismatched (i.e. `ast` is not the result of parsing `statement`). I slightly changed the wording in the latest commit. Let me know if this should be changed further